Query by hash prefix

Zegnat opened the pull request:
As discussed in #25 this allows a client to request a list of segments based on on the first bytes of a hashed video ID. Thus not having to send every video ID to the SponsorBlockServer.

Marking this PR as a work-in-progress until tests have been added to confirm that the implementation does what it is supposed to do.

It is based on the `getSkipSegments` function but with some big differences:

1. The logic around `shadowHidden` has been redone. I _think_ this leads to the same results. The query to the privateDB has been moved outside of the for-loop and the check has been made much more straight forward (IMO). Hopefully this can speed up the response time.
2. The logic in `getVoteOrganisedSponsorTimes` is not yet implemented. I will look into implementing that. But as the new set of segments includes multiple videos it needs more wrangling than the previous implementation was capable of.

## API:
### GET `/api/skipSegments/:prefix`
Replace `:prefix` with between 2 and 32 hexadecimal characters (default) making up the start of the hashed video ID.
#### Response:
```jsonc
[{ // Array of this object
   videoID: string,
   segment: float[], //[0, 15.23] start and end time in seconds
   UUID: string,
   category: string
}]
```
#### Error codes:
<dl><dt>400: Bad Request</dt><dd>When the prefix provided was outside of the range supported by the server.</dd></dl>

ajayyy commented on databases/_upgrade_sponsorTimes_2.sql:
```diff
@@ -0,0 +1,26 @@
+	"views"	INTEGER NOT NULL,
+	"category"	TEXT NOT NULL DEFAULT "sponsor",
+	"shadowHidden"	INTEGER NOT NULL,
+	"hashedVideoID"	TEXT NOT NULL
```
Maybe put this right before shadow hidden and leave shadow hidden as the last column

Zegnat replied:
I don’t think anything is SQLite depends on column order? But I’ve swapped them in https://github.com/ajayyy/SponsorBlockServer/pull/86/commits/3167c24f75bcd9c390baff39c72a47130e2d16f8.

ajayyy replied:
It just looks better when opening the database in a gui viewer. Thanks!

ajayyy commented on src/routes/getSkipSegmentsByHash.js:
```diff
@@ -0,0 +1,69 @@
+ * @param {string} hashedIP Custom hash of the visitor’s IP address
+ * @returns {Segment[]}
+ */
+function getSkipSegmentsByHash(prefix, hashedIP) {
```
Hmm, I'm not sure about this. Is `category` really a large privacy concern?

Zegnat replied:
No. I just didn’t know if it even made sense to split by category at all? Maybe it does and I should add that as a choice to the URL.

ajayyy replied:
Segments from different categories may overlap, and would end up being filtered out by the `getVoteOrganisedSponsorTimes` logic. So, it has to be split by category.

ajayyy commented on src/routes/getSkipSegmentsByHash.js:
```diff
@@ -0,0 +1,69 @@
+        getHash(getIP(req) + config.globalSalt)
+    );
+
+    if (segments) {
+        res.send(segments)
+    }
```
else, send error

Zegnat replied:
Added better check and error response in https://github.com/ajayyy/SponsorBlockServer/pull/86/commits/c7fd603933e7d33661c7d3ba7fb0486411b619b4.

ajayyy commented on test/cases/getSavedTimeForUser.js:
```diff
@@ -5,8 +5,8 @@ var getHash = require('../../src/utils/getHash.js');
-    let startOfQuery = "INSERT INTO sponsorTimes (videoID, startTime, endTime, votes, UUID, userID, timeSubmitted, views, category, shadowHidden) VALUES";
-    db.exec(startOfQuery + "('getSavedTimeForUser', 1, 11, 2, 'abc1239999', '" + getHash("testman") + "', 0, 50, 'sponsor', 0)");
+    let startOfQuery = "INSERT INTO sponsorTimes (videoID, startTime, endTime, votes, UUID, userID, timeSubmitted, views, category, shadowHidden, hashedVideoID) VALUES";
+    db.exec(startOfQuery + "('getSavedTimeForUser', 1, 11, 2, 'abc1239999', '" + getHash("testman") + "', 0, 50, 'sponsor', 0, '" + getHash('getSavedTimeForUser', 0) + "')");
```
`getSavedTimeForUser` is confusing for a videoID. Maybe choose something more obivous.

Zegnat replied:
I am just hashing the name that was already in the test data. Someone that came before me picked `getSavedTimeForUser` as the videoID value for the database row.

Should I change nonsensical videoID values like this one whenever I see them in tests? What about those that include things like commas?

ajayyy replied:
Oh sorry, I didn't notice this was actually a test for `getSavedTimeForUser`, I guess I understand why the name is like that. You can leave it.

Joe-Dowd replied:
> Should I change nonsensical videoID values like this one whenever I see them in tests? What about those that include things like commas?
The commas could be to test to make sure the application is able to deal with commas in the data (a , isn't a valid character in a video id but _ and - are for example)

Zegnat commented:
In https://github.com/ajayyy/SponsorBlockServer/pull/86/commits/c2510d302a3934a8c895c03b943403721b6df763 I implemented the `getVoteOrganisedSponsorTimes` logic. Overlapping segments are grouped and then a weighted random pick is made from each group.

The only thing left to implement from that is the logic that [limits each video to only 8 segments](https://github.com/ajayyy/SponsorBlockServer/blob/51f760332bb157730185d4118b081c4cb355bad9/src/routes/getSkipSegments.js#L180-L183).

I have also changed the return body. It no longer returns a single array, instead it returns an object with a member per video. This should make it easier for a consuming client to find the segments it was actually interested in. Rather than having to loop the entire collection it can instantly get the collection it wants.

## Updated API:
### GET `/api/skipSegments/:prefix`
Replace `:prefix` with between 2 and 32 hexadecimal characters (default) making up the start of the hashed video ID.
#### Response:
```jsonc
{ // Object of all found videos, multiple members like:
  "videoID": [{ // Array of segments
    videoID: string,
    segment: float[], //[0, 15.23] start and end time in seconds
    UUID: string,
    category: string
  }]
}
```
#### Error codes:
<dl><dt>400: Bad Request</dt><dd>When the prefix provided was outside of the range supported by the server.</dd></dl>

ajayyy commented:
I think `videoID` can be removed from the element if it is now the key. Great idea on that.

ajayyy commented:
I'd prefer if you use url parameters to be consistent with the rest of the API

Zegnat commented:
> I think videoID can be removed from the element if it is now the key.

Agreed.

> I'd prefer if you use url parameters to be consistent with the rest of the API

Sure. This was mostly modelled after the Pwned Password API. The theoretical good thing with it is that the entire API can be provided by just static files. But as SponsorBlockServer has randomisation and hidden records I guess that isn’t going to apply here either way.

What parameters do we imagine working well here? I was thinking `prefix` and `category`? By making the queries per category it gives a simple out from [the previous conversation](https://github.com/ajayyy/SponsorBlockServer/pull/86#discussion_r429562844).

ajayyy commented:
> Sure. This was mostly modelled after the Pwned Password API. The theoretical good thing with it is that the entire API can be provided by just static files. But as SponsorBlockServer has randomisation and hidden records I guess that isn’t going to apply here either way.

I see. Maybe that can be done to the (future) raw api endpoint.

> What parameters do we imagine working well here? I was thinking prefix and category? By making the queries per category it gives a simple out from the previous conversation.

Sounds good. Ideally it should be like the original and allow you to get multiple categories at once as well.

Joe-Dowd commented on src/routes/postSkipSegments.js:
```diff
@@ -232,9 +232,9 @@ module.exports = async function postSkipSegments(req, res) {
-                    segmentInfo.segment[1], startingVotes, UUID, userID, timeSubmitted, 0, segmentInfo.category, shadowBanned);
+                    "(videoID, startTime, endTime, votes, UUID, userID, timeSubmitted, views, category, shadowHidden, hashedVideoID)" +
+                    "VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)").run(videoID, segmentInfo.segment[0],
+                    segmentInfo.segment[1], startingVotes, UUID, userID, timeSubmitted, 0, segmentInfo.category, shadowBanned, getHash(videoID, 0));
```
`getHash(videoID, 0)` will result in no passes on the string for hashing (returning an empty string to protect data that should be hashed). Given the DB migration script, it should be 1.
https://github.com/ajayyy/SponsorBlockServer/blob/master/src/utils/getHash.js#L4

Recommend creating some tests that check the hashing of the video id is being done correctly on insert.

Likewise all the uses in the test scripts need to be changes.

ajayyy replied:
The test scripts should probably be hardcoded instead of being dependent on `getHash`

Joe-Dowd commented:
Also this should be merged into testing, not master

Joe-Dowd commented on src/routes/getSkipSegmentsByHash.js:
```diff
@@ -0,0 +1,125 @@
+
+const minimumPrefix = config.minimumPrefix || '3';
+const maximumPrefix = config.maximumPrefix || '32'; // Half the hash.
+const prefixChecker = new RegExp('^[\\dA-F]{' + minimumPrefix + ',' + maximumPrefix + '}$', 'i');
```
Regex checks for upper case, but prefix is converted to lowercase before being sent to getSkipSegmentsByHash - is there any point enforcing case when submitting a hash? If so i'd personally side on submitting in the same form as it's stored in the DB.

Joe-Dowd commented:
I just pulled the PR to try to find a prefix length that gave a reasonable amount of returned data, but noticed the migration script isn't working:
```
sqlite> select hashedVideoID from sponsorTimes limit 3;
0
0
0
sqlite>
```

Joe-Dowd commented:
I wrote a script to take a prefix length and output the amount of videos that will be returned for a prefix (taken from existing video id hashes):

```#!/bin/bash
set -e


PREFIX_LENGTH="${1:-4}"

echo "Prefix Length: ${PREFIX_LENGTH} (length of string is -1 due to sqlite substring)"
echo ''
echo 'number of occurances, distinct video ids, number of videos effected'

sqlite3 sponsorTimes.db "select count(distinct videoId), substr(hashedVideoId, 0, ${PREFIX_LENGTH}) from sponsorTimes where hashedVideoID LIKE substr(hashedVideoId, 0, ${PREFIX_LENGTH}) || '%' group by substr(hashedVideoId, 0, ${PREFIX_LENGTH})" | awk -F '|' '{print $1}' | sort -n | uniq -c | awk '{sep= ","; print $1, sep, $2, sep, $1 * $2;}' | sed 's/ ,/,/g'```

and imo the length that we should be using is 3 characters:

```
Prefix Length: 4 (length of string is -1 due to sqlite substring)

number of occurances, distinct video ids, number of videos effected
2, 7, 14
1, 8, 8
6, 9, 54
24, 10, 240
25, 11, 275
51, 12, 612
69, 13, 897
117, 14, 1638
151, 15, 2265
234, 16, 3744
259, 17, 4403
325, 18, 5850
362, 19, 6878
341, 20, 6820
373, 21, 7833
341, 22, 7502
305, 23, 7015
251, 24, 6024
218, 25, 5450
177, 26, 4602
148, 27, 3996
95, 28, 2660
70, 29, 2030
50, 30, 1500
36, 31, 1116
22, 32, 704
22, 33, 726
12, 34, 408
3, 35, 105
4, 36, 144
1, 38, 38
1, 43, 43
```

ajayyy commented:
This was probably caused by me telling him to swap the columns in the upgrade script. He probably just forgot to swap it in that spot.

girst commented:
I had a superficial look at this PR, and I think it looks pretty good. I'm a bit concerned by the duplication of logic in getSkipSegmentsByHash(), but that could be cleaned up after a merge.

girst commented:
@Zegnat: how's this coming along? anything you're stuck on?

Zegnat commented:
The only blocker for me  right now has been _time_. I hope to get around back to this PR soon!

girst commented:
On Thu, Jun 18, 2020 at 09:42:36AM -0700, Martijn van der Ven wrote:
>The only blocker for me  right now has been _time_. I hope to get around back to this PR soon!

no worries! i appreciate you tackling this.

Joe-Dowd commented:
https://github.com/ajayyy/SponsorBlockServer/pull/127
